### PR TITLE
[AC-1923] Add endpoint to create client organization

### PR DESCRIPTION
**Pull request by amorask-bitwarden**
## Type of change

```
- [ ] Bug fix
- [X] New feature development
- [X] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

The goal of this PR was to allow for the creation of client organizations in the Consolidated Billing workflow. It adds a new billing-owned endpoint at `POST /providers/{providerId}/clients` that takes in some information about a client organization and then does the following:
1. Performs some Admin Console functions using the `ProviderService.CreateOrganizationAsync` method. (When the organization is created, it's given 0 seats to start by design).
2. Assigns seats to the client organization so we can re-use the command that autoscales the provider's subscription.
3. Creates a Stripe customer for the client organization in the event they ever unlink from the MSP.

I've left in-PR comments annotating the changes that I think would be easier to follow than listing out all the files below.

Companion `clients` PR: https://github.com/bitwarden/clients/pull/8748

https://github.com/bitwarden/server/assets/144709477/abca11c4-0d6d-4ef7-853c-c107273b3960

### CODE REVIEW

File: bitwarden_license/src/Commercial.Core/AdminConsole/Services/ProviderService.cs [hunk 1]
```diff
@@ -38,13 +38,14 @@ public class ProviderService : IProviderService
     private readonly IOrganizationService _organizationService;
     private readonly ICurrentContext _currentContext;
     private readonly IStripeAdapter _stripeAdapter;
+    private readonly IFeatureService _featureService;
 
     public ProviderService(IProviderRepository providerRepository, IProviderUserRepository providerUserRepository,
```
Comment (amorask-bitwarden) [inline bitwarden_license/src/Commercial.Core/AdminConsole/Services/ProviderService.cs:43]: In the `ProviderService`, I updated the `CreateOrganizationAsync` method to pivot off the `enable-consolidated-billing` feature flag. When it's On, it invokes the `OrganizationService.SignupClientAsync` instead of `SignupAsync`.

File: src/Api/Billing/Controllers/ProviderBillingController.cs [hunk 2]
```diff
@@ -28,17 +28,17 @@ public async Task<IResult> GetSubscriptionAsync([FromRoute] Guid providerId)
             return TypedResults.Unauthorized();
         }
 
-        var subscriptionData = await providerBillingQueries.GetSubscriptionData(providerId);
+        var providerSubscriptionDTO = await providerBillingQueries.GetSubscriptionDTO(providerId);
```
Comment (amorask-bitwarden) [inline src/Api/Billing/Controllers/ProviderBillingController.cs:31]: Everything in this file is renames to better organize how we refer to specific types of models.

File: src/Api/Billing/Controllers/ProviderClientsController.cs [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+using Bit.Core.Context;
+using Bit.Core.Models.Business;
+using Bit.Core.Repositories;
+using Bit.Core.Services;
+using Microsoft.AspNetCore.Mvc;
+
+namespace Bit.Api.Billing.Controllers;
+
+[Route("providers/{providerId:guid}/clients")]
+public class ProviderClientsController(
```
Comment (amorask-bitwarden) [inline src/Api/Billing/Controllers/ProviderClientsController.cs:15]: This file is a rename from `ProviderOrganizationsController` to emphasize it's specific to Consolidated Billing. The only net-new portion is the `CreateAsync` endpoint which will be used to create a new client organization under an MSP.

File: src/Api/Billing/Models/Requests/CreateClientOrganizationRequestBody.cs [hunk 4]
```diff
@@ -0,0 +1,29 @@
+﻿using System.ComponentModel.DataAnnotations;
+using Bit.Api.Utilities;
+using Bit.Core.Enums;
+
+namespace Bit.Api.Billing.Models.Requests;
+
+public class CreateClientOrganizationRequestBody
```
Comment (amorask-bitwarden) [inline src/Api/Billing/Models/Requests/CreateClientOrganizationRequestBody.cs:7]: A representation of the request sent by the client to create a new client organization. We need significantly less information than we do when spinning up a traditional organization.

File: src/Api/Billing/Models/Requests/UpdateClientOrganizationRequestBody.cs [hunk 5]
```diff
@@ -0,0 +1,10 @@
+﻿using System.ComponentModel.DataAnnotations;
+
+namespace Bit.Api.Billing.Models.Requests;
+
+public class UpdateClientOrganizationRequestBody
```
Comment (amorask-bitwarden) [inline src/Api/Billing/Models/Requests/UpdateClientOrganizationRequestBody.cs:5]: This is a rename from `UpdateProviderOrganizationRequestBody` to not confuse it with AC's version. I also added the data annotation attributes, but it is not related to the ticket's acceptance criteria.

File: src/Api/Billing/Models/Responses/ProviderSubscriptionResponse.cs [hunk 6]
```diff
@@ -2,9 +2,9 @@
 using Bit.Core.Utilities;
 using Stripe;
 
-namespace Bit.Api.Billing.Models;
+namespace Bit.Api.Billing.Models.Responses;
 
-public record ProviderSubscriptionDTO(
```
Comment (amorask-bitwarden) [inline src/Api/Billing/Models/Responses/ProviderSubscriptionResponse.cs:7]: Unrelated renaming for consistency.

File: src/Api/Utilities/EnumMatchesAttribute.cs [hunk 7]
```diff
@@ -0,0 +1,26 @@
+﻿using System.ComponentModel.DataAnnotations;
+
+namespace Bit.Api.Utilities;
+
+public class EnumMatchesAttribute<T>(params T[] accepted) : ValidationAttribute
```
Comment (amorask-bitwarden) [inline src/Api/Utilities/EnumMatchesAttribute.cs:5]: I added this attribute to validate an incoming `PlanType`, but it can be used for any enum.

File: src/Core/AdminConsole/Services/Implementations/OrganizationService.cs [hunk 8]
```diff
@@ -421,6 +421,81 @@ public async Task VerifyBankAsync(Guid organizationId, int amount1, int amount2)
         }
     }
 
+    public async Task<(Organization organization, OrganizationUser organizationUser, Collection defaultCollection)> SignupClientAsync(OrganizationSignup signup)
```
Comment (amorask-bitwarden) [inline src/Core/AdminConsole/Services/Implementations/OrganizationService.cs:424]: This is a newly added method responsible for signing up a client organization for an MSP as part of Consolidated Billing.

File: src/Core/Billing/Commands/IAssignSeatsToClientOrganizationCommand.cs [hunk 9]
```diff
@@ -5,6 +5,15 @@ namespace Bit.Core.Billing.Commands;
 
 public interface IAssignSeatsToClientOrganizationCommand
```
Comment (amorask-bitwarden) [inline src/Core/Billing/Commands/IAssignSeatsToClientOrganizationCommand.cs:6]: 🔧 Added missing documentation.

File: src/Core/Billing/Models/ConfiguredProviderPlanDTO.cs [hunk 10]
```diff
@@ -3,17 +3,17 @@
 
 namespace Bit.Core.Billing.Models;
 
-public record ConfiguredProviderPlan(
```
Comment (amorask-bitwarden) [inline src/Core/Billing/Models/ConfiguredProviderPlanDTO.cs:6]: 🔧 Rename for consistency.

File: src/Core/Billing/Models/ProviderSubscriptionDTO.cs [hunk 11]
```diff
@@ -0,0 +1,7 @@
+﻿using Stripe;
+
+namespace Bit.Core.Billing.Models;
+
+public record ProviderSubscriptionDTO(
```
Comment (amorask-bitwarden) [inline src/Core/Billing/Models/ProviderSubscriptionDTO.cs:5]: 🔧 Rename from `ProviderSubscriptionData` for consistency.

File: src/Core/Billing/Queries/ISubscriberQueries.cs [hunk 12]
```diff
@@ -6,6 +6,18 @@ namespace Bit.Core.Billing.Queries;
 
 public interface ISubscriberQueries
```
Comment (amorask-bitwarden) [inline src/Core/Billing/Queries/ISubscriberQueries.cs:7]: Added two new methods to safely retrieve a Stripe `Customer` using a `ISubscriber`. This is used in the `CreateCustomerCommand`.

File: bitwarden_license/src/Commercial.Core/AdminConsole/Services/ProviderService.cs [hunk 13]
```diff
@@ -611,8 +619,13 @@ private async Task<bool> HasConfirmedProviderAdminExceptAsync(Guid providerId, I
         return confirmedOwnersIds.Except(providerUserIds).Any();
     }
 
-    private void ThrowOnInvalidPlanType(PlanType requestedType)
+    private void ThrowOnInvalidPlanType(PlanType requestedType, bool consolidatedBillingEnabled)
```
Comment (vincentsalucci) [inline bitwarden_license/src/Commercial.Core/AdminConsole/Services/ProviderService.cs:622]: Non-blocking suggestion: could just default this parameter to `false` to simplify extra code changes (ex: in the `AddOrganization` method above)

File: src/Core/AdminConsole/Services/Implementations/OrganizationService.cs (hunk 8 above)
Comment (vincentsalucci) [reply src/Core/AdminConsole/Services/Implementations/OrganizationService.cs:424]: Non-blocking comment: Hmmm. Yet another fork on creating an organization. @eliykat Any thoughts on how we mitigate/improve this fragmentation?

File: src/Core/AdminConsole/Enums/OrganizationStatusType.cs [hunk 14]
```diff
@@ -3,5 +3,6 @@
 public enum OrganizationStatusType : byte
 {
     Pending = 0,
-    Created = 1
+    Created = 1,
+    Managed = 2,
```
Comment (amorask-bitwarden) [inline src/Core/AdminConsole/Enums/OrganizationStatusType.cs:7]: This is set when an organization is fully brought under an MSP in Consolidated Billing.

File: src/Core/AdminConsole/Services/Implementations/OrganizationService.cs (hunk 8 above)
Comment (amorask-bitwarden) [reply src/Core/AdminConsole/Services/Implementations/OrganizationService.cs:424]: @vincentsalucci @eliykat Just to emphasize, this is relatively peculiar situation given we're creating organizations with paid features that do not have any payment information. That's all managed at the provider level so I need to be able to skip that process.

File: bitwarden_license/src/Commercial.Core/AdminConsole/Services/ProviderService.cs (hunk 13 above)
Comment (amorask-bitwarden) [reply bitwarden_license/src/Commercial.Core/AdminConsole/Services/ProviderService.cs]: Added! https://github.com/bitwarden/server/pull/3977/commits/051e1418cc2078c8f61b037b493ed7e273c72178

File: src/Core/AdminConsole/Services/Implementations/OrganizationService.cs (hunk 8 above)
Comment (eliykat) [reply src/Core/AdminConsole/Services/Implementations/OrganizationService.cs:424]: I have a tech debt task to move org signup methods into their own command: https://bitwarden.atlassian.net/browse/AC-1517

My vision here is a `CreateOrganizationCommand` with different public methods for different create flows (via web app, via provider portal, via Bitwarden portal, etc...)

I don't like adding to `OrganizationService` but I think it may as well live here with the other signup methods so that they're all together for us to move later. So I'm OK with this PR, we just gotta continue breaking up OrganizationService (AC Team work).

**Comment by amorask-bitwarden** [comment]
Hi @vincentsalucci, given @eliykat's earlier comment, did you need anything else for the review? Thanks!

**Review by cturnbull-bitwarden — APPROVED** [review]

**Review by vincentsalucci — APPROVED** [review]
